### PR TITLE
[DEV-797] - Add multi-give function and forge tests

### PR DESCRIPTION
**Pull request by grp06**
## Describe your changes
In this PR I add a function `multiGive` that allows a badge spec holder to pass in an array of wallet addresses and mint multiple badges at once. 

To stay in compliance with ERC-4973 I left the `give` function alone and created `multiGive` (open to a different name). I had to create a new function since 1. I couldn't directly modify the arguments of `give` to support "multi give" without breaking compliance with ERC-4973 and  2. `give` is external, so I can't call it from the contract

## Gas costs
- multigive
    - 3 recipients - 307215
    - 2 recipients - 207575
    - 1 recipient - 107938
- give
    - 1 recipient 106855

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have updated the documentation accordingly. (if applicable)
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/Badges.sol [hunk 1]
```diff
@@ -115,6 +115,23 @@ contract Badges is
     specDataHolder = ISpecDataHolder(_dataHolder);
   }
 
+  function multiGive(
```
Comment (rahulrumalla) [inline src/Badges.sol:118]: Rename to `giveToMany`

**Review by rahulrumalla — CHANGES_REQUESTED** [review]
Couple of minor things

File: src/Badges.sol [hunk 2]
```diff
@@ -115,6 +115,23 @@ contract Badges is
...
+    string calldata _uri,
+    bytes[] calldata _signature
+  ) external {
+    uint256 raftTokenId = specDataHolder.getRaftTokenId(_uri);
+    require(
+      specDataHolder.getRaftOwner(raftTokenId) == msg.sender,
+      "give: unauthorized"
+    );
+    for (uint256 i = 0; i < _recipients.length; i++) {
+      require(msg.sender != _recipients[i], "give: cannot give to self");
```
[the remark this answers is not in the document]
Comment (grp06) [reply src/Badges.sol]: done

File: src/Badges.t.sol [hunk 3]
```diff
@@ -921,4 +921,81 @@ contract BadgesTest is Test {
     (uint8 v, bytes32 r, bytes32 s) = vm.sign(raftHolderPrivateKey, hash);
     signature = abi.encodePacked(r, s, v);
   }
+
```
[the remark this answers is not in the document]
Comment (grp06) [reply src/Badges.t.sol:924]: Just wrote a test for it. If one fails they all fail

File: src/Badges.sol [hunk 4]
```diff
@@ -115,6 +115,44 @@ contract Badges is
     specDataHolder = ISpecDataHolder(_dataHolder);
   }
 
+  function _give(
+    address _recipient,
+    string calldata _uri,
+    bytes calldata _signature,
+    uint256 _raftTokenId
+  ) internal returns (uint256) {
```
Comment (rahulrumalla) [inline src/Badges.sol:123]: should this be virtual?

File: src/Badges.sol [hunk 5]
```diff
@@ -115,6 +115,44 @@ contract Badges is
...
+   * @notice Allows the owner of a badge spec to mint a badge to multiple recipeients who have requested it
+   * @param _recipients array the addresses who will receive a badge
+   * @param _uri the uri of the badge spec
+   * @param _signatures array of signatures that verify that the person receiving the badge actually requested it
+   */
+  function giveToMany(
+    address[] calldata _recipients,
+    string calldata _uri,
+    bytes[] calldata _signatures
+  ) external {
```
Comment (rahulrumalla) [inline src/Badges.sol:139]: virtual here too?

File: src/Badges.t.sol [hunk 6]
```diff
@@ -921,4 +924,166 @@ contract BadgesTest is Test {
...
+      active,
+      recipient2PrivateKey
+    );
+
+    bytes memory recipient3Signature = getSignature(
+      active,
+      recipient3PrivateKey
+    );
+
+    address[] memory recipientsAddresses = new address[](3);
```
Comment (rahulrumalla) [inline src/Badges.t.sol:954]: could we refactor this as this is used across tests?

File: src/Badges.t.sol [hunk 7]
```diff
@@ -921,4 +924,166 @@ contract BadgesTest is Test {
...
+    recipientsAddresses[2] = recipient3;
+
+    bytes[] memory recipientsSignatures = new bytes[](3);
+    recipientsSignatures[0] = recipient1Signature;
+    recipientsSignatures[1] = recipient2Signature;
+    recipientsSignatures[2] = recipient3Signature;
+
+    vm.prank(active);
+    // expect it to revert since we passed in a bad signature
+    vm.expectRevert(bytes(errInvalidSig));
```
Comment (rahulrumalla) [inline src/Badges.t.sol:1010]: Here one issue is that when 1 of 100 entries fail, how do we let the caller know it was item "65" that actually failed?

File: src/Badges.sol [hunk 8]
```diff
@@ -115,6 +115,44 @@ contract Badges is
...
+      "giveToMany: recipients and signatures length mismatch"
+    );
+
+    uint256 raftTokenId = specDataHolder.getRaftTokenId(_uri);
+    require(
+      specDataHolder.getRaftOwner(raftTokenId) == msg.sender,
+      "giveToMany: unauthorized"
+    );
+
+    for (uint256 i = 0; i < _recipients.length; i++) {
```
Comment (rahulrumalla) [inline src/Badges.sol:151]: gas optimization opportunity

https://hackmd.io/@totomanov/gas-optimization-loops

**Review by rahulrumalla — CHANGES_REQUESTED** [review]

File: src/Badges.sol (hunk 8 above)
Comment (grp06) [reply src/Badges.sol:151]: For minting 3 badges, putting the for loop inside of an `unchecked` takes `giveToMany` from 306213 to 306013 ... I'd say we hold off on `unchecked` for now until I research it further and feel more confident with it. (we can always upgrade)

**Review by rahulrumalla — APPROVED** [review]

**Pull request merged**